### PR TITLE
Expanded docs for `deprecation_date`

### LINKED ISSUE #3745: Expanded `deprecation_date` docs: motivation, examples, tie-in to model governance
### Contributions

- [X] I have read the contribution docs, and understand what's expected of me.

### Link to the page on docs.getdbt.com requiring updates

https://docs.getdbt.com/reference/resource-properties/deprecation_date

- Link back to page providing thorough overview of model versions
- What does "YAML DateTime format" mean? What are the viable options?
- Add an example, including 

https://docs.getdbt.com/docs/collaborate/govern/model-versions

These statements need updating, and should link back to the reference docs on `deprecation_date`:

> In the future, dbt will also offer first-class support for deprecating models (https://github.com/dbt-labs/dbt-core/issues/7433).
> 
> dbt will notify consumers of your versioned model when new versions become available, or (in the future) when they are slated for deprecation
> 
> Because dbt knows that these models are actually the same model, it can notify downstream consumers as new versions become available, and (in the future) as older versions are slated for deprecation.

### What part(s) of the page would you like to see updated?

https://github.com/dbt-labs/docs.getdbt.com/pull/3579 added the stub for the `deprecation_date` page, but we're missing a lot of important context & detail:
- why would I set a `deprecation_date` for my models?
- what happens after a model is deprecated? does it stop building/being selected? what are the warning/error messages I could expect to see, for myself and for downstream queriers?
- how does this relate to model versions? do I have to be using model contracts/versions/etc to set a `deprecation_date`? (no!)
- how does this relate to [table expiration on BQ](https://docs.getdbt.com/reference/resource-configs/bigquery-configs#controlling-table-expiration)? (they're different things, but maybe you'd want to use them in synchrony)
- is there specific selection syntax I can use, to select only deprecated models? (I don't think we've implemented this)
- how can I clean up / remove tables from the DWH associated with deprecated models? (nothing built-in/automated, but is there a discourse/pattern we could point people toward?)

### Additional information

- https://github.com/dbt-labs/dbt-core/pull/7562
- https://github.com/dbt-labs/dbt-core/issues/7433

### PR DESCRIPTION
**Pull request by dbeatty10**
## What are you changing in this pull request and why?

Resolves #3745

- Updated docs for [model versions](https://docs.getdbt.com/docs/collaborate/govern/model-versions) including link back to the reference docs on `deprecation_date`
- Add examples with different accepted formats, including an aware datetime, a naive datetime, and a plain date
   - `1999-01-01 00:00:00.00+00:00`
   - `2023-12-31 23:59:59.99`
   - `2024-01-01`

## Previews
- [deprecation_date](https://deploy-preview-3754--docs-getdbt-com.netlify.app/reference/resource-properties/deprecation_date)
- [model-versions](https://deploy-preview-3754--docs-getdbt-com.netlify.app/docs/collaborate/govern/model-versions)

## 🎩 

image

### Why would I set a `deprecation_date` for my models?

Declaring a deprecation date is a way of signaling the maturity level of a model, and communicating plans for long-term support and maintenance. It provides a mechanism to communicate plans and timelines for sunsetting models so they don't need to be maintained and supported indefinitely and their build and storage costs can be freed up.

### What happens after a model is deprecated?

Does it stop building/being selected? What are the warning/error messages I could expect to see, for myself and for downstream queriers?

Deprecated models can still continue to be built by producers and be selected by consumers until they are [disabled](https://docs.getdbt.com/reference/resource-configs/enabled) or removed.

Models marked for deprecation will raise informative warnings when they are parsed or referenced. (With cross-project references, dbt raises these warnings in downstream projects that are ref'ing a deprecated version of a public model from an upstream project.)

When a project references a model that's slated for deprecation, a warning is generated. If it's a versioned model, with a newer version available, then the warning says so. This added bit of communication, from producers to consumers, is an advantage of using dbt's built-in functionality around model versions to facilitate migrations.

Additionally, [`WARN_ERROR_OPTIONS`](https://docs.getdbt.com/reference/global-configs/warnings) gives a mechanism whereby users can promote these warnings to actual runtime errors.

**Warning messages:**
  * DeprecatedModel (warning when parsing a project that defines deprecated model(s)): `[WARNING] Model {model_name}[.v{version} if versioned] has passed its deprecation date of {deprecation_date}. This model should be disabled or removed.`
  * UpcomingDeprecationReference (warning when referencing a model with a future deprecation date): `[WARNING] While compiling '{this_model_name}': Found a reference to {ref_model_name}[.v{version} if versioned], which is slated for deprecation on '{ref_model_deprecation_date}'. [if versioned: A new version of '{ref_model_name}' is available. Try it out: {{ ref('ref_model_package', 'ref_model_name', v='{ref_model_latest_version}') }}.`
  * DeprecatedReference (warning when referencing a model with a past deprecation date): `[WARNING] While compiling '{this_model_name}': Found a reference to {ref_model_name}[.v{version} if versioned], which was deprecated on '{ref_model_deprecation_date}'. [if versioned: A new version of '{ref_model_name}' is available. Migrate now: {{ ref('ref_model_package', 'ref_model_name', v='{ref_model_latest_version}') }}.`

### How does this relate to model versions?

Do I have to be using model contracts/versions/etc to set a `deprecation_date`?

(no!)

Setting a `deprecation_date` works well in conjunction with other [model governance](https://docs.getdbt.com/docs/collaborate/govern/about-model-governance) features like [model versions](https://docs.getdbt.com/docs/collaborate/govern/model-versions), but can also be used independently from them.

### Is there specific selection syntax I can use, to select only deprecated models?

(I don't think we've implemented this)

There is not specific [node selection syntax](https://docs.getdbt.com/reference/node-selection/syntax) for `deprecation_date`. [Programmatic invocations](https://docs.getdbt.com/reference/programmatic-invocations) is one way to identify deprecated models (potentially in conjunction with [dbt list](https://docs.getdbt.com/reference/commands/list)). e.g., `dbt -q ls  --output json --output-keys database schema alias deprecation_date`.

### How can I clean up / remove tables from the DWH associated with deprecated models?

(nothing built-in/automated, but is there a discourse/pattern we could point people toward?)

Just like it [does not automatically drop relations when models are deleted](https://docs.getdbt.com/faqs/models/removing-deleted-models), dbt does not removed tables for deprecated models.

Strategies similar to [here](https://discourse.getdbt.com/t/faq-cleaning-up-removed-models-from-your-production-schema/113) or [here](https://discourse.getdbt.com/t/clean-your-warehouse-of-old-and-deprecated-models/1547) can be used to drop relations that have been deprecated and are no longer in use.

### How does this relate to [table expiration on BQ](https://docs.getdbt.com/reference/resource-configs/bigquery-configs#controlling-table-expiration)?

(they're different things, but maybe you'd want to use them in synchrony)

dbt-bigquery can set an [`hours_to_expiration`](https://docs.getdbt.com/reference/resource-configs/bigquery-configs#controlling-table-expiration) that [translates to `expiration_timestamp`](https://github.com/dbt-labs/dbt-bigquery/blob/ea258bb76169375ded8f7ff9e596a436a5ed165a/dbt/adapters/bigquery/impl.py#L844-L846) (expired tables in BigQuery will be deleted and their storage reclaimed).

dbt does not automatically synchronize `deprecation_date` and `hours_to_expiration`, but users may want to coordinate them in some fashion (such as setting a model to expire 48 hours after its `deprecation_date`).

## Checklist
- [x] Review the [Content style guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md) and [About versioning](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version) so my content adheres to these guidelines.

### CODE REVIEW

File: website/docs/reference/resource-properties/deprecation_date.md [hunk 1]
```diff
@@ -28,5 +28,46 @@ models:
...
+
+### Purpose
+
+Declaring a `deprecation_date` for a dbt model provides a mechanism to communicate plans and timelines for long-term support and maintenance and to facilitate change management.
+
+Setting a `deprecation_date` works well in conjunction with other [model governance](/docs/collaborate/govern/about-model-governance) features like [model versions](/docs/collaborate/govern/model-versions), but can also be used independently from them.
+
+### Warning messages
+
+When a project references a model that's slated for deprecation or the deprecation date has passed, a warning is generated. If it's a versioned model, with a newer version available, then the warning says so. This added bit of cross-team communication, from producers to consumers, is an advantage of using dbt's built-in functionality around model versions to facilitate migrations.
```
Comment (jtcohen6) [inline website/docs/reference/resource-properties/deprecation_date.md:48]: 👍 love this

Could we include a code snippet displaying the terminal/IDE output for one of these warnings? I think `UpcomingDeprecationReference` is the most compelling

File: website/docs/reference/resource-properties/deprecation_date.md [hunk 2]
```diff
@@ -28,5 +28,46 @@ models:
...
+
+### Deprecation process
+
+Additional steps are necessary to save on build-related compute and storage costs for a deprecated model.
+
+Deprecated models can continue to be built by producers and be selected by consumers until they are [disabled](/reference/resource-configs/enabled) or removed.
+
+Just like it does not automatically [drop relations when models are deleted](/faqs/models/removing-deleted-models), dbt does not drop relations for deprecated models.
+
+Strategies similar to [here](https://discourse.getdbt.com/t/faq-cleaning-up-removed-models-from-your-production-schema/113) or [here](https://discourse.getdbt.com/t/clean-your-warehouse-of-old-and-deprecated-models/1547) can be used to drop relations that have been deprecated and are no longer in use.
```
Comment (jtcohen6) [inline website/docs/reference/resource-properties/deprecation_date.md:80]: good links!

File: website/docs/reference/resource-properties/deprecation_date.md [hunk 3]
```diff
@@ -28,5 +28,46 @@ models:
...
+
+### Table expiration on BigQuery
+
+dbt-bigquery can set an [`hours_to_expiration`](/reference/resource-configs/bigquery-configs#controlling-table-expiration) that translates to `expiration_timestamp` within BigQuery.
+
+dbt does not automatically synchronize `deprecation_date` and `hours_to_expiration`, but users may want to coordinate them in some fashion (such as setting a model to expire 48 hours after its `deprecation_date`). Expired tables in BigQuery will be deleted and their storage reclaimed.
```
Comment (jtcohen6) [inline website/docs/reference/resource-properties/deprecation_date.md:86]: > setting a model to expire 48 hours after its `deprecation_date`

And it's reasonable enough to do this, using the `datetime` module? That's neat!

File: website/docs/reference/resource-properties/deprecation_date.md [hunk 4]
```diff
@@ -28,5 +28,46 @@ models:
...
+Setting a `deprecation_date` works well in conjunction with other [model governance](/docs/collaborate/govern/about-model-governance) features like [model versions](/docs/collaborate/govern/model-versions), but can also be used independently from them.
+
+### Warning messages
+
+When a project references a model that's slated for deprecation or the deprecation date has passed, a warning is generated. If it's a versioned model, with a newer version available, then the warning says so. This added bit of cross-team communication, from producers to consumers, is an advantage of using dbt's built-in functionality around model versions to facilitate migrations.
+
+Additionally, [`WARN_ERROR_OPTIONS`](/reference/global-configs/warnings) gives a mechanism whereby users can promote these warnings to actual runtime errors:
+- `DeprecatedModel` - warning when parsing a project that defines deprecated model(s)
+- `UpcomingDeprecationReference` - warning when referencing a model with a future deprecation date
+- `DeprecatedReference` - warning when referencing a model with a past deprecation date
```
Comment (jtcohen6) [inline website/docs/reference/resource-properties/deprecation_date.md]: Maybe some clarification that:
- `DeprecatedModel` is for the _producer_ of the deprecated model
- `UpcomingDeprecationReference` and `DeprecatedReference` are for the _consumers_ of the deprecated model (could be in a different project!)

**Review by jtcohen6 — APPROVED** [review]
Thank you @dbeatty10!! This is a great set of updates :)

Tiny comments, not blockers for merge.

I'm also realizing we implemented `deprecated_date` as a model property, not config, so it won't be possible for users to configure a `deprecation_date` for many models at once from `dbt_project.yml` (like `access`). Another thing to revisit as part of a larger refactor or consolidation of configs/properties.

File: website/docs/reference/resource-properties/deprecation_date.md (hunk 1 above)
Comment (dbeatty10) [reply website/docs/reference/resource-properties/deprecation_date.md:48]: Addressed in 5ec8a46fbe98c253995f3b878ca158f296963e21

File: website/docs/reference/resource-properties/deprecation_date.md (hunk 3 above)
Comment (dbeatty10) [reply website/docs/reference/resource-properties/deprecation_date.md:86]: To unblock merging this PR, deferring the actual demonstration of this to a separate issue: #3765

**Pull request merged**